### PR TITLE
feat: Add Tenderly Simulation before relaying

### PR DESCRIPTION
**Pull request by yagopv**
## Summary

This pull request introduces a new feature that enables pre-relay transaction simulation using Tenderly for selected chain IDs. The main goal is to block relay attempts if a simulation predicts the transaction would fail

## Changes

**Tenderly Simulation Integration**

* Added `ITenderlySimulationApi` interface and result types in `src/domain/interfaces/tenderly-simulation-api.interface.ts` to define how simulations are performed and how results are reported.
* Implemented the `TenderlySimulationApi` service (with dependency injection module) to handle simulation logic, including error handling, log analysis for `ExecutionFailure`, and gas limit caching.
* Introduced schema validation for Tenderly simulation responses using Zod in `src/modules/relay/datasources/schemas/tenderly-simulation.schema.ts`, with corresponding unit tests for edge cases and validation logic. [[1]](diffhunk://#diff-f3e2369d33bddb1a45ca569fd8817b2969ae185c074587abcae1b54d951f4de3R1-R26) [[2]](diffhunk://#diff-3e67b303524f0afe0a05c1393c573e2053b4b4f4697164292a7258a58404645cR1-R95)

**Configuration Changes**

* Added `RELAY_SIMULATION_CHAIN_IDS` to `.env.sample.json` to allow configuration of which chain IDs require pre-simulation, and updated the dynamic configuration code to parse and expose this value. [[1]](diffhunk://#diff-ab7dd5bc4f296032b95e861acb01d599f30e41cd04a5e8a1378145286e96858fR1184-R1189) [[2]](diffhunk://#diff-a13ec7a4d04b743754a74ef5f390ded5f6b6aeb41e68de735adba4f15b9df7edR706-R710) [[3]](diffhunk://#diff-8e9fd4781f9f2ef0d5ba9d9c955b9d74d993588ca787837d4da9e103dd3373d0R401-R403)

**Testing**

* Added a comprehensive test suite for the `TenderlySimulationApi` service, covering success, failure, error handling, log topic analysis, and caching behavior.

### CODE REVIEW

File: src/modules/relay/datasources/tenderly-simulation-api.service.ts [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+      } else if (error instanceof NetworkRequestError) {
+        const cause = error.data;
+        const causeStr =
+          cause instanceof Error
+            ? `${cause.name}: ${cause.message}${
+                'cause' in cause && cause.cause
+                  ? ` (cause: ${cause.cause instanceof Error ? `${cause.cause.name}: ${cause.cause.message}` : String(cause.cause)})`
+                  : ''
+              }${cause.stack ? `\n${cause.stack}` : ''}`
+            : JSON.stringify(cause);
```
Comment (compojoom) [inline src/modules/relay/datasources/tenderly-simulation-api.service.ts]: this is very hard to read. Let's move it to some utility with few ifs.

**Comment by compojoom** [comment]
@claude review

File: src/modules/relay/datasources/tenderly-simulation-api.service.ts [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+        url: SIMULATION_URL,
+        data: {
+          network_id: args.chainId,
+          from: args.from,
+          to: args.to,
+          input: args.data,
+          value: (args.value ?? 0n).toString(),
+          gas,
+          gas_price: '0',
+          save: true,
```
Comment (compojoom) [inline src/modules/relay/datasources/tenderly-simulation-api.service.ts]: isn't save going to persist every simulation?

Comment (yagopv) [reply src/modules/relay/datasources/tenderly-simulation-api.service.ts]: I maintained the web behavior, but i guess with save_if_fails should be enough to debug, right?

Comment (yagopv) [reply src/modules/relay/datasources/tenderly-simulation-api.service.ts]: Updated to only save_if_fails

File: src/modules/relay/datasources/tenderly-simulation-api.service.ts (hunk 1 above)
Comment (yagopv) [reply src/modules/relay/datasources/tenderly-simulation-api.service.ts]: done

**Comment by yagopv** [comment]
@claude review

File: src/modules/relay/datasources/tenderly-simulation-api.service.ts [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+    data: Hex;
+    value?: bigint;
+  }): Promise<TenderlySimulationResult> {
+    try {
+      // Always use the latest block gas limit as the simulation budget; the
+      // relay request's `gasLimit` is what Gelato will use on-chain and is
+      // typically too tight to cover the simulation's overhead (refund path,
+      // etc.). Mirrors what the Safe frontend does when no explicit
+      // simulation gas budget is supplied.
+      const gas = Number(await this.getLatestBlockGasLimit(args.chainId));
```
Comment (dasanra) [inline src/modules/relay/datasources/tenderly-simulation-api.service.ts:147]: Is it necessary to convert a BigInt to a Number?

**Review by dasanra — COMMENTED** [review]
I think swagger is not updated with this PR changes

File: src/modules/relay/datasources/tenderly-simulation-api.service.ts [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+  type ILoggingService,
+  LoggingService,
+} from '@/logging/logging.interface';
+import { TenderlySimulationResponseSchema } from '@/modules/relay/datasources/schemas/tenderly-simulation.schema';
+
+/**
+ * Public Tenderly proxy used by the Safe frontend. No auth required and the
+ * response is the standard Tenderly simulation payload.
+ */
+const SIMULATION_URL = 'https://simulation.safe.global/';
```
Comment (LucieFaire) [inline src/modules/relay/datasources/tenderly-simulation-api.service.ts:33]: to make sure, is it the same one for staging and prod?

File: src/modules/relay/datasources/tenderly-simulation-api.service.ts [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+    value?: bigint;
+  }): Promise<TenderlySimulationResult> {
+    try {
+      // Always use the latest block gas limit as the simulation budget; the
+      // relay request's `gasLimit` is what Gelato will use on-chain and is
+      // typically too tight to cover the simulation's overhead (refund path,
+      // etc.). Mirrors what the Safe frontend does when no explicit
+      // simulation gas budget is supplied.
+      const gas = Number(await this.getLatestBlockGasLimit(args.chainId));
+      const { data: raw } = await this.networkService.post<unknown>({
```
Comment (LucieFaire) [inline src/modules/relay/datasources/tenderly-simulation-api.service.ts]: anyway we can avoid `unknown` type here?

File: src/modules/relay/datasources/tenderly-simulation-api.service.ts [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+function formatCause(cause: unknown): string {
+  return cause instanceof Error
+    ? `${cause.name}: ${cause.message}`
+    : String(cause);
+}
+
+function formatError(error: Error): string {
+  const stack = error.stack ? `\n${error.stack}` : '';
+  return `${error.name}: ${error.message}${stack}`;
+}
```
Comment (LucieFaire) [inline src/modules/relay/datasources/tenderly-simulation-api.service.ts:69]: could we extract it to some utils file?

File: src/modules/relay/domain/relayers/relay-fee.relayer.ts [hunk 7]
```diff
@@ -117,8 +138,10 @@ export class RelayFeeRelayer implements IRelayer {
       await this.validateExecTransaction({
         chainId,
         to,
+        data,
         decoded,
         safeTxHash: args.safeTxHash,
+        acceptUnverifiedSimulation: args.acceptUnverifiedSimulation,
```
Comment (LucieFaire) [inline src/modules/relay/domain/relayers/relay-fee.relayer.ts]: nit: maybe we can also destruct those two from args on L126?

**Review by LucieFaire — DISMISSED** [review]
left some nice to haves, but not blockers

File: src/modules/relay/domain/relayers/__tests__/relay-fee.relayer.spec.ts [hunk 8]
```diff
@@ -406,6 +417,221 @@ describe('RelayFeeRelayer', () => {
...
+        expect(result).toEqual({ taskId });
+        expect(mockRelayApi.relay).toHaveBeenCalledTimes(1);
+        expect(mockLoggingService.warn).toHaveBeenCalledWith(
+          expect.objectContaining({
+            message: expect.stringContaining('user override'),
+          }),
+        );
+      });
+
+      it('still throws RelaySimulationFailedError on the override path when the second simulation confirms a revert', async () => {
```
Comment (dasanra) [inline src/modules/relay/domain/relayers/__tests__/relay-fee.relayer.spec.ts]:
```suggestion
      it('throws RelaySimulationFailedError whith acceptUnverifiedSimulation not overriding a confirmed simulation failure', async () => {
```

File: src/modules/relay/datasources/tenderly-simulation-api.service.ts (hunk 3 above)
Comment (yagopv) [reply src/modules/relay/datasources/tenderly-simulation-api.service.ts:147]: Yes, I tried String(...)  and simulation.safe.global rejects it with:
```
HTTP 400 json_unmarshal: Bad request input parameters
```

the proxy enforces number for gas. The lossy bigint → Number cast is safe in practice

File: src/modules/relay/datasources/tenderly-simulation-api.service.ts (hunk 4 above)
Comment (yagopv) [reply src/modules/relay/datasources/tenderly-simulation-api.service.ts:33]: yep

File: src/modules/relay/datasources/tenderly-simulation-api.service.ts (hunk 5 above)
Comment (yagopv) [reply src/modules/relay/datasources/tenderly-simulation-api.service.ts]: done

File: src/modules/relay/domain/relayers/relay-fee.relayer.ts (hunk 7 above)
Comment (yagopv) [reply src/modules/relay/domain/relayers/relay-fee.relayer.ts]: done

**Review by LucieFaire — APPROVED** [review]

**Review by dasanra — APPROVED** [review]
